### PR TITLE
Added authentication for restricted access articles

### LINKED ISSUE #438: Solutions for restricted access articles
When an article is on a website which restricts access (most presumably because you are paying for it), several possibilities could be used to fetch content : 
- enter credentials and send them as a http header for sites which are compatible
- copy/paste content in wallabag
- save original article as a .html file and then import it in wallabag
- use a browser extension and send content via JS

However, all of this should be implemented as plugins in v2, not directly into Wallabag.

What do you think ?

### PR DESCRIPTION
**Pull request by nicosomb**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Documentation | no |
| Translation | no |
| Fixed tickets | #438 |
| License | MIT |

Fix #438.
Previously at #1586
Thank you so much @bdunogier

Todo 
- [x] Documentation
- [ ] Move websites login/password in the database
- [x] Tests

### CODE REVIEW

**Comment by nicosomb** [comment]
Tested with mediapart.fr and arretsurimages.net ✅ 

About arretsurimages.net, here is the config site file: 

```
title://div[@id="titrage-contenu"]/h1[@class="title"]
body: //div[@class="contenu-html"]/div[@class="page-pane"]

requires_login: yes

login_uri: http://www.arretsurimages.net/forum/login.php
login_username_field: username
login_password_field: password

not_logged_in_xpath: //body[@class="not-logged-in"]
``` 

I tried for nextinpact.com but it fails. I think it's because the login form is not in the HTML DOM (it's loaded in ajax).

**Comment by bdunogier** [comment]
Can I take this over somehow ? I know I've been silent for a while, but I still care for it :-)

In any case, I'm willing to help with finishing this up.

@nicosomb what did you change from the original PR ?

**Comment by nicosomb** [comment]
Nothing. I just rebased and opened a new PR. 

I tested your work :+1:

Happy to see that you're back!

**Comment by bdunogier** [comment]
I'm never far, I'm just overloaded ;-)

> Nothing. I just rebased and opened a new PR.
Oh, my. For some reason, I didn't remember writing the tests.

In that case, I'd appreciate if you could set the author. It's a small thing, but well, you know ? Or I could open a new PR if there are things you'd like me to change before merging.

About things to do, regarding NextINpact, I guess that the fix should go into Graby, since that's where the config is. I'll start having a look.

**Comment by nicosomb** [comment]
> In that case, I'd appreciate if you could set the author of the commit.

no problem, I will do that.

**Comment by nicosomb** [comment]
@bdunogier done.

**Comment by nicosomb** [comment]
@bdunogier this feature is planned for 2.2, which will be released on december. 
Can you tell us if you will find time to finish this PR?

File: app/Resources/CraueConfigBundle/translations/CraueConfigBundle.en.yml [hunk 1]
```diff
@@ -32,3 +32,4 @@ demo_mode_enabled: "Enable demo mode ? (only used for the wallabag public demo)"
 demo_mode_username: "Demo user"
 share_public: Allow public url for entries
 download_images_enabled: Download images locally
+restricted_access: Enable authentication for websites with paywall
```
Comment (j0k3r) [inline app/Resources/CraueConfigBundle/translations/CraueConfigBundle.en.yml]: Do not forget other translations files.

File: src/Wallabag/CoreBundle/Helper/HttpClientFactory.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    private $cookieJar;
+
+    private $restrictedAccess;
+
+    /**
+     * HttpClientFactory constructor.
+     *
+     * @param \GuzzleHttp\Event\SubscriberInterface $authenticatorSubscriber
+     * @param \GuzzleHttp\Cookie\CookieJar          $cookieJar
+     * @param string                                $restrictedAccess
```
Comment (j0k3r) [inline src/Wallabag/CoreBundle/Helper/HttpClientFactory.php]: Maybe you can precise that, even if's a string, it should be a kind of boolean.

File: app/Resources/CraueConfigBundle/translations/CraueConfigBundle.en.yml (hunk 1 above)
Comment (nicosomb) [inline app/Resources/CraueConfigBundle/translations/CraueConfigBundle.en.yml]: Oops!

**Comment by j0k3r** [comment]
Don't forget to run php-cs-fixer

**Comment by nicosomb** [comment]
:tada:

**Comment by fradav** [comment]
I don't know if it is the my config, the authentication module or the graby recipe, but it doesn't work for mediapart.fr authenticated access. I already noticed there is a "div" for the body in the recipe which has to be changed to "main".

**Comment by nicosomb** [comment]
Hello @fradav ! 

Can you open a new issue here please https://github.com/wallabag/wallabag/issues/new
with details of your issue.

**Pull request merged**